### PR TITLE
Integrate tps_performance_monitor into trx_generator program

### PR DESCRIPTION
**Pull request by ndcgundlach**
Modified trx_generator to add parameters for tps_performance_monitor.  Added creation of tps_performance_monitor in main for trx_generator.

### CODE REVIEW

File: tests/trx_generator/main.cpp [hunk 1]
```diff
@@ -48,6 +52,9 @@ int main(int argc, char** argv) {
          ("trx-gen-duration", bpo::value<uint32_t>(&gen_duration)->default_value(60), "Transaction generation duration (seconds). Defaults to 60 seconds.")
          ("target-tps", bpo::value<uint32_t>(&target_tps)->default_value(1), "Target transactions per second to generate/send. Defaults to 1 transaction per second.")
          ("last-irreversible-block-id", bpo::value<string>(&lib_id_str), "Current last-irreversible-block-id (LIB ID) to use for transactions.")
+         ("monitor-spinup-time-us", bpo::value<int64_t>(&spinup_time_us)->default_value(1000000), "Number of microseconds to wait before monitoring TPS. Defaults to 1000000 (1s).")
+         ("monitor-max-lag-percent", bpo::value<uint32_t>(&max_lag_per)->default_value(0), "Max percentage off from expected transactions sent before being in violation. Defaults to 5.")
+         ("monitor-max-lag-duration-us", bpo::value<int64_t>(&max_lag_duration_us), "Max microseconds that transaction generation can be in violation before quitting. Defaults to 1000000 (1s).")
```
Comment (oschwaldp-oci) [inline tests/trx_generator/main.cpp:57]: Does not look like the default was applied here.

File: tests/trx_generator/main.cpp [hunk 2]
```diff
@@ -48,6 +52,9 @@ int main(int argc, char** argv) {
          ("trx-gen-duration", bpo::value<uint32_t>(&gen_duration)->default_value(60), "Transaction generation duration (seconds). Defaults to 60 seconds.")
          ("target-tps", bpo::value<uint32_t>(&target_tps)->default_value(1), "Target transactions per second to generate/send. Defaults to 1 transaction per second.")
          ("last-irreversible-block-id", bpo::value<string>(&lib_id_str), "Current last-irreversible-block-id (LIB ID) to use for transactions.")
+         ("monitor-spinup-time-us", bpo::value<int64_t>(&spinup_time_us)->default_value(1000000), "Number of microseconds to wait before monitoring TPS. Defaults to 1000000 (1s).")
+         ("monitor-max-lag-percent", bpo::value<uint32_t>(&max_lag_per)->default_value(0), "Max percentage off from expected transactions sent before being in violation. Defaults to 5.")
```
Comment (oschwaldp-oci) [inline tests/trx_generator/main.cpp:56]: Says defaults to 5 but default value is 0.

File: tests/trx_generator/main.cpp [hunk 3]
```diff
@@ -112,6 +119,22 @@ int main(int argc, char** argv) {
...
       }
+
+      if(vmap.count("spinup-time-us")) {
+         if(spinup_time_us < 0) {
+            ilog("Initialization error: spinup-time-us cannot be negative");
+            cli.print(std::cerr);
+            return INITIALIZE_FAIL;
+         }
+      }
+
```
Comment (oschwaldp-oci) [inline tests/trx_generator/main.cpp:130]: Should a check be added for "monitor-max-lag-percent" being in a reasonable % range?  At the very least check it is in percentage range 0-100?

**Review by oschwaldp-oci — CHANGES_REQUESTED** [review]

File: tests/trx_generator/main.cpp (hunk 2 above)
Comment (ndcgundlach) [reply tests/trx_generator/main.cpp]: fixed

File: tests/trx_generator/main.cpp (hunk 3 above)
Comment (ndcgundlach) [reply tests/trx_generator/main.cpp:130]: added validation

**Review by oschwaldp-oci — APPROVED** [review]

**Pull request merged**